# **STUDENT AI** - DATA CLEANING

## Objectives

Inspect the dataset and solve any issues that might arise from wrong data types, or missing / wrong values

## Inputs

Continues to assess dataset loaded in previous notebook.

## Outputs

Saves the cleaned dataset back to inputs/dataset folder for futher use


---

# Import required libraries

In [1]:
import os
import pandas as pd
from pandas_profiling import ProfileReport

print('All Libraries Loaded')

All Libraries Loaded


/tmp/ipykernel_8135/881926697.py:3: DeprecationWarning: `import pandas_profiling` is going to be deprecated by April 1st. Please use `import ydata_profiling` instead.
  from pandas_profiling import ProfileReport


# Change working directory

### Set the working directory to notebook parent folder
If the output does not match, click **'clear all outputs'** and then **'restart'** the notebook. 
Then run cells from top to bottom.

In [2]:
current_dir = os.getcwd()
os.chdir(os.path.dirname(current_dir))
current_dir = os.getcwd()
print('If correct, Active Directory should read: /workspace/student-AI')
print(f"Active Directory: {current_dir}")

If correct, Active Directory should read: /workspace/student-AI
Active Directory: /workspace/student-AI


### Load saved dataset

In [3]:
df = pd.read_csv(f"inputs/dataset/Expanded_data_with_more_features.csv")
df.head()

Unnamed: 0  Gender EthnicGroup          ParentEduc     LunchType TestPrep  \
0           0  female         NaN   bachelor's degree      standard     none   
1           1  female     group C        some college      standard      NaN   
2           2  female     group B     master's degree      standard     none   
3           3    male     group A  associate's degree  free/reduced     none   
4           4    male     group C        some college      standard     none   

  ParentMaritalStatus PracticeSport IsFirstChild  NrSiblings TransportMeans  \
0             married     regularly          yes         3.0     school_bus   
1             married     sometimes          yes         0.0            NaN   
2              single     sometimes          yes         4.0     school_bus   
3             married         never           no         1.0            NaN   
4             married     sometimes          yes         0.0     school_bus   

  WklyStudyHours  MathScore  ReadingScore  WritingScore  
0            < 5         71            71            74  
1         5 - 10         69            90            88  
2            < 5         87            93            91  
3         5 - 10         45            56            42  
4         5 - 10         76            78            75

### Drop Unnamed column that pandas created on import

In [4]:
df.drop(columns=['Unnamed: 0'], inplace=True)
df.head()

Gender EthnicGroup          ParentEduc     LunchType TestPrep  \
0  female         NaN   bachelor's degree      standard     none   
1  female     group C        some college      standard      NaN   
2  female     group B     master's degree      standard     none   
3    male     group A  associate's degree  free/reduced     none   
4    male     group C        some college      standard     none   

  ParentMaritalStatus PracticeSport IsFirstChild  NrSiblings TransportMeans  \
0             married     regularly          yes         3.0     school_bus   
1             married     sometimes          yes         0.0            NaN   
2              single     sometimes          yes         4.0     school_bus   
3             married         never           no         1.0            NaN   
4             married     sometimes          yes         0.0     school_bus   

  WklyStudyHours  MathScore  ReadingScore  WritingScore  
0            < 5         71            71            74  
1         5 - 10         69            90            88  
2            < 5         87            93            91  
3         5 - 10         45            56            42  
4         5 - 10         76            78            75

### Summarize Dataset Again

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30641 entries, 0 to 30640
Data columns (total 14 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Gender               30641 non-null  object 
 1   EthnicGroup          28801 non-null  object 
 2   ParentEduc           28796 non-null  object 
 3   LunchType            30641 non-null  object 
 4   TestPrep             28811 non-null  object 
 5   ParentMaritalStatus  29451 non-null  object 
 6   PracticeSport        30010 non-null  object 
 7   IsFirstChild         29737 non-null  object 
 8   NrSiblings           29069 non-null  float64
 9   TransportMeans       27507 non-null  object 
 10  WklyStudyHours       29686 non-null  object 
 11  MathScore            30641 non-null  int64  
 12  ReadingScore         30641 non-null  int64  
 13  WritingScore         30641 non-null  int64  
dtypes: float64(1), int64(3), object(10)
memory usage: 3.3+ MB


At first glance, 30000+ data rows is a very robust dataset to be able to train an ML. <br>
There seems to be no reason why the NrSiblings should be a float data type as you cannot have 0.25 of a brother/sister
we can convert the column to an integer type - however this throws an error as the dataset seems to have NaN or Inf data, so first we deal with missing data

### Check for missing values

In [6]:
missing_values = df.isnull().sum()
percentage_missing = (missing_values / len(df)) * 100
percentage_missing = percentage_missing.round(1)
missing_data = pd.DataFrame({'Missing_Values': missing_values, 'Percentage_Missing': percentage_missing})
missing_data['Percentage_Missing'] = missing_data['Percentage_Missing'].astype(str) + '%'
missing_data

Missing_Values Percentage_Missing
Gender                            0               0.0%
EthnicGroup                    1840               6.0%
ParentEduc                     1845               6.0%
LunchType                         0               0.0%
TestPrep                       1830               6.0%
ParentMaritalStatus            1190               3.9%
PracticeSport                   631               2.1%
IsFirstChild                    904               3.0%
NrSiblings                     1572               5.1%
TransportMeans                 3134              10.2%
WklyStudyHours                  955               3.1%
MathScore                         0               0.0%
ReadingScore                      0               0.0%
WritingScore                      0               0.0%

### Assess how many rows would need to be dropped...

In [7]:
dropped_data = df.dropna()
dropped_data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 19243 entries, 2 to 30640
Data columns (total 14 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Gender               19243 non-null  object 
 1   EthnicGroup          19243 non-null  object 
 2   ParentEduc           19243 non-null  object 
 3   LunchType            19243 non-null  object 
 4   TestPrep             19243 non-null  object 
 5   ParentMaritalStatus  19243 non-null  object 
 6   PracticeSport        19243 non-null  object 
 7   IsFirstChild         19243 non-null  object 
 8   NrSiblings           19243 non-null  float64
 9   TransportMeans       19243 non-null  object 
 10  WklyStudyHours       19243 non-null  object 
 11  MathScore            19243 non-null  int64  
 12  ReadingScore         19243 non-null  int64  
 13  WritingScore         19243 non-null  int64  
dtypes: float64(1), int64(3), object(10)
memory usage: 2.2+ MB


## Initial Assumptions

* Dropping null values leaves us with 19243 data rows ... I hypothesize that while this seems like enough data to still achieve the business requirement, dropping this many values will likely induce an imbalance to the data and bias the dataset. If confirmed, then imputing logical values to fill the gaps is more advisable.
* The Pandas report shows **Alerts** under the nrSiblings variable as it assumes that a value of zero (0) could be a problem. In this case, it can be ignored, as a value of zero (0) is valid and indicates an only child.


## Assessing imbalance on 'dropped' dataset

In [ ]:
pandas_report = ProfileReport(df=dropped_data, minimal=True)
pandas_report.to_notebook_iframe()

## Pandas Report on 'dropped' dataset

Clicking on each variable to view the specific report fo each variable:

| Variable |  Obvious Insights  | 
|---|---|
| Gender  |  shows equal disribution of 50.8% to 49.2% and no missing values |
|EthnicGroup|report is not much value as it gives the word 'group' most emphasis - **string 'group ' needs to be removed from data**|
|ParentEduc|again here the word 'some' is not accurate. this categorical variable should be changed to a linear numerical one indicating 0 none to n highest level of education, with 'some as intermediary values|
|LunchType|this feature is imbalanced with 64.8% standard and 35.2% free/reduced. Will require some feature engineering|
|TestPrep|this feature is imbalanced with 65.4% completed and 34.6% none. Will require some feature engineering|
|ParentMaritalStatus|this feature is imbalanced with the majority rows (57.2%) labelled married|
|practiceSport|this feature is imbalanced with the majority rows (50.5%) labelled sometimes will need to be engineered or dropped|
|IsFirstChild|this feature is imbalanced with 64.5% yes and 35.6% no. Will require some feature engineering|
|NrSiblings|this numerical feature looks skewed and has a few outliers that might be able to be removed - further analysis required|
|TransportMeans|this feature is slightly imbalanced 58.6% private vs 41.4% school_bus|
|WklyStudyHours|this is a categorical feature with options <5, 5-10 ,and >10 hours ... distribution is reasonable balanced 39.3%/32.4%/28.3%|
|MathScore|at first glance looks like normal distribution, but obvious missing values that were dropped|
|ReadingScore|at first glance looks like normal distribution, but more obvious missing values that were dropped|
|WritingScore|at first glance looks like normal distribution, but some missing values that were dropped|


### dropping the missing value data rows has indeed induced an imbalance to the dataset. I will therefore assess the viability of imputing data to fill the gaps...

## Imputing Missing Values
let's review the missing_data form the full dataset

In [8]:
missing_data

Missing_Values Percentage_Missing
Gender                            0               0.0%
EthnicGroup                    1840               6.0%
ParentEduc                     1845               6.0%
LunchType                         0               0.0%
TestPrep                       1830               6.0%
ParentMaritalStatus            1190               3.9%
PracticeSport                   631               2.1%
IsFirstChild                    904               3.0%
NrSiblings                     1572               5.1%
TransportMeans                 3134              10.2%
WklyStudyHours                  955               3.1%
MathScore                         0               0.0%
ReadingScore                      0               0.0%
WritingScore                      0               0.0%

The highest missing data rate is TransportMeans at 10.2%. Thinking logically about how this feature might affect overall school performance of a student, it seems that this is an indicator of economic status of the family.
Another indicator of economic status of the family is LunchType, indicating whether the family has the means available to pay for own school lunches. This feature has no missing values.

I will therefore **drop the TransportMeans** feature as it has many missing values and a single indicator of Family economic status should be sufficient in assessing its effect on school performance.

In [9]:
df.drop(columns=['TransportMeans'], inplace=True)

### Imputing Categorical Variables

Categorical variables with missing values are :
* 'EthnicGroup'
* 'TestPrep'
* 'ParentEduc'
* 'ParentMaritalStatus'
* 'IsFirstChild'
* 'PracticeSport'
* 'WklyStudyHours'
* ('LunchType' and 'Gender' have no missing values and do not need to be adjusted)

For these categorical variables I will insert the most common value from the dataset (mode) as that will be closest to the actual value probabalistically.

In [10]:
for column in ['EthnicGroup', 'TestPrep', 'ParentEduc', 'ParentMaritalStatus', 'IsFirstChild', 'PracticeSport', 'WklyStudyHours']:
    mode_value = df[column].mode()[0]
    df[column].fillna(mode_value, inplace=True)

### Imputing Numerical Variables

The only numerical variable from the feature set (not counting the scores which have no missing values) is NrSiblings.
Once the missing values have been imputed, I can also change the data type to a more sensical integer rather than float.
The imputed values will be based on the **median** instead of the **mean** as this is less sensitive to outliers since the variable does contain some 'extreme' values of 6 or more siblings.

In [11]:
median_value = df['NrSiblings'].median()
df['NrSiblings'] = df['NrSiblings'].fillna(median_value).astype(int)

### Quick check for remaining missing values and check datatype change and possible duplicate values:

In [12]:
df.isnull().sum()

Gender                 0
EthnicGroup            0
ParentEduc             0
LunchType              0
TestPrep               0
ParentMaritalStatus    0
PracticeSport          0
IsFirstChild           0
NrSiblings             0
WklyStudyHours         0
MathScore              0
ReadingScore           0
WritingScore           0
dtype: int64

In [13]:
df['NrSiblings'].dtype

dtype('int64')

### The data is now clean and logical values have been added, in the next book I will conduct an EDA to go into detail about the feature set and data distribution / balance. 

## Save file to repository for follow on notebooks

In [ ]:
file_path = 'outputs/dataset/Expanded_data_with_more_features_clean.csv'

# Remove previous file if it exists
if os.path.exists(file_path):
    os.remove(file_path)

# Create the directory if it doesn't exist
os.makedirs(name='outputs/dataset', exist_ok=True)

# Save cleaned DataFrame to the file path
df.to_csv(file_path, index=False)